### PR TITLE
[IR] Add per-function numbers to basic blocks

### PR DESCRIPTION
**Pull request by aengelke**
Every basic block that is linked into a function now has a unique number, which can be queried using getNumber(). Numbers are densely allocated, but not re-assigned on block removal for stability. Block numbers are intended to be fairly stable and only be updated when removing a several basic blocks to make sure the numbering doesn't become too sparse.

To reduce holes in the numbering, renumberBlocks() can be called to re-assign numbers in block order.

Previous discussion in: https://discourse.llvm.org/t/rfc-add-auxiliary-field-for-per-pass-custom-data-to-basicblock/80229

---

I added a validate method to catch cases where something goes wrong, even if I can't really imagine how invalid numbers can occur. But I think it's better to be safe and rule out this potential source of bugs when more things depend on the numbering.

Not sure if I should add a `getNumBlockIDs()` that returns NextBlockNum. So far, I haven't felt the need. I think we can always add it later if it turns out to be beneficial.

### CODE REVIEW

File: llvm/include/llvm/IR/Function.h [hunk 1]
```diff
@@ -996,6 +1012,12 @@ class LLVM_EXTERNAL_VISIBILITY Function : public GlobalObject,
   void setValueSubclassDataBit(unsigned Bit, bool On);
 };
 
+#ifdef NDEBUG
+/// In release builds, this is a no-op. For !NDEBUG builds, the checks are
+/// implemented in the .cpp file.
+inline void Function::validateBlockNumbers() const {}
+#endif
```
Comment (nikic) [inline llvm/include/llvm/IR/Function.h:1019]: I'd prefer not to have the conditional definitions. Especially as the only callers are currently in the same file, there is no penalty to having the ifdef NDEBUG over the function body only.

File: llvm/include/llvm/IR/Function.h [hunk 2]
```diff
@@ -75,6 +75,22 @@ class LLVM_EXTERNAL_VISIBILITY Function : public GlobalObject,
...
+
+public:
+  /// Renumber basic blocks into a dense value range starting from 0. Be aware
+  /// that other data structures and analyses (e.g., DominatorTree) may depend
+  /// on the value numbers and need to be updated or invalidated.
+  void renumberBlocks();
+
+  /// Assert that all blocks have unique numbers within 0..NextBlockNum. This
+  /// has O(n) runtime complexity.
+  void validateBlockNumbers() const;
```
Comment (nikic) [inline llvm/include/llvm/IR/Function.h:91]: Can make this private, or do you expect this to be called from more places in future patches?

File: llvm/unittests/IR/FunctionTest.cpp [hunk 3]
```diff
@@ -487,6 +487,78 @@ TEST(FunctionTest, EraseBBs) {
   EXPECT_EQ(F->size(), 0u);
 }
 
+TEST(FunctionTest, BasicBlockNumbers) {
+  LLVMContext Context;
+  Type *VoidType(Type::getVoidTy(Context));
+  FunctionType *FuncType(FunctionType::get(VoidType, false));
```
Comment (nikic) [inline llvm/unittests/IR/FunctionTest.cpp:493]:
```suggestion
  Type *VoidType = Type::getVoidTy(Context);
  FunctionType *FuncType = FunctionType::get(VoidType, false);
```

File: llvm/include/llvm/IR/Function.h (hunk 1 above)
Comment (aengelke) [reply llvm/include/llvm/IR/Function.h]: Agreed, changed

**Comment by aengelke** [comment]
Changed. I also decided to add getMaxBlockNumber() here. I think this will come useful.

Furthermore, I introduced the concept of "block number epoch" -- an integer that changes for every renumbering. This will be useful (=already was for me) for identifying use of block numbers after renumbering: on initialization, the current epoch is stored, and on all subsequent accesses, equality with the current epoch can be asserted.

File: llvm/include/llvm/IR/Function.h [hunk 4]
```diff
@@ -75,6 +75,38 @@ class LLVM_EXTERNAL_VISIBILITY Function : public GlobalObject,
...
   // Important things that make up a function!
   BasicBlockListType BasicBlocks;         ///< The basic blocks
+
+  // Basic blocks need to get their number when added to a function.
+  friend void BasicBlock::setParent(Function *);
+  unsigned NextBlockNum = 0;
+  /// Epoch of block numbers. (Could be shrinked to uint8_t if required.)
+  unsigned BlockNumEpoch = 0;
+
+public:
```
Comment (nikic) [inline llvm/include/llvm/IR/Function.h:85]: Can you please move these functions after all the variable declarations? It makes it hard to see what the layout of the structure is if everything is interleaved.

**Review by nikic — APPROVED** [review]
LGTM

**Pull request merged**